go/oasis-test-runner/scenario/e2e/runtime/runtime_upgrade.go

```diff
@@ -0,0 +1,276 @@
+	}
+
+	kmRuntimeEncID := kmRuntime.GetEnclaveIdentity()
+	oldRuntimeEncID := oldRuntime.GetEnclaveIdentity()
+	newRuntimeEncID := newRuntime.GetEnclaveIdentity()
+
+	if oldRuntimeEncID == nil && newRuntimeEncID == nil {
+		sc.Logger.Info("No SGX runtimes, skipping policy update")
+		return nil
+	}
+
+	// Ensure enclave IDs differ between the old and new runtimes.
+	oldEncID, _ := oldRuntimeEncID.MarshalText()
+	newEncID, _ := newRuntimeEncID.MarshalText()
+	if bytes.Equal(oldEncID, newEncID) {
+		return fmt.Errorf("expected different enclave identities, got: %s", newEncID)
+	}
+
+	// Build updated SGX policies.
+	sc.Logger.Info("building new KM SGX policy enclave policies map")
+	enclavePolicies := make(map[sgx.EnclaveIdentity]*keymanager.EnclavePolicySGX)
+
+	enclavePolicies[*kmRuntimeEncID] = &keymanager.EnclavePolicySGX{}
+	enclavePolicies[*kmRuntimeEncID].MayQuery = make(map[common.Namespace][]sgx.EnclaveIdentity)
+
+	// Allow new compute runtimes to query private data.
+	for _, rt := range sc.Net.Runtimes() {
+		if rt.Kind() != registry.KindCompute {
+			continue
+		}
+		if eid := rt.GetEnclaveIdentity(); eid != nil {
+			enclavePolicies[*kmRuntimeEncID].MayQuery[rt.ID()] = []sgx.EnclaveIdentity{*eid}
+		}
+	}
+
+	sc.Logger.Info("initing updated KM policy")
+	if err := cli.Keymanager.InitPolicy(kmRuntime.ID(), 2, enclavePolicies, kmPolicyPath); err != nil {
+		return err
+	}
+	sc.Logger.Info("signing updated KM policy")
+	if err := cli.Keymanager.SignPolicy("1", kmPolicyPath, kmPolicySig1Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("2", kmPolicyPath, kmPolicySig2Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("3", kmPolicyPath, kmPolicySig3Path); err != nil {
+		return err
+	}
+
+	sc.Logger.Info("updating KM policy")
+	if err := cli.Keymanager.GenUpdate(sc.nonce, kmPolicyPath, []string{kmPolicySig1Path, kmPolicySig2Path, kmPolicySig3Path}, kmUpdateTxPath); err != nil {
+		return err
+	}
+	if err := cli.Consensus.SubmitTx(kmUpdateTxPath); err != nil {
+		return fmt.Errorf("failed to update KM policy: %w", err)
+	}
+	sc.nonce++
+
+	return nil
+}
+
+func (sc *runtimeUpgradeImpl) Run(childEnv *env.Env) error {
+	ctx := context.Background()
+	cli := cli.New(childEnv, sc.Net, sc.Logger)
+
+	computeWorkers := len(sc.Net.ComputeWorkers())
+
+	clientErrCh, cmd, err := sc.runtimeImpl.start(childEnv)
+	if err != nil {
+		return err
+	}
+	sc.Logger.Info("waiting for client to exit")
+	// Wait for the client to exit.
+	select {
+	case err = <-sc.runtimeImpl.Net.Errors():
+		_ = cmd.Process.Kill()
+	case err = <-clientErrCh:
+	}
+	if err != nil {
+		return err
+	}
+
+	// Generate and update a policy that will allow the new runtime to run.
+	if err = sc.applyUpgradePolicy(childEnv); err != nil {
+		return fmt.Errorf("updating policies: %w", err)
+	}
+
+	// Kill old compute workers.
+	sc.Logger.Info("stopping old runtimes")
+	for i := 0; i < computeWorkers/2; i++ {
```

Comment:
Instead of using `computeWorkers / 2` would it be better to have an index where new workers start?